### DISCUSSION TITLE
Sync Imap account

### DISCUSSION THREAD

**Discussion by GP-S** [twentyhq/twenty#918] [Ideas]
Good feature available in hubspot for instance is the ability to synchronize your email account and attach mails or conversation from a customer to an opportunity/company account. 

It can be down by adding a kind of inbox for all messages and some auto filters on the sender and recipient field of the mail.

Main idea is to store discussion/attached file directly in the crm without need to copy paste your mail conversation. 

P.S: Of course it can be extended to linkedin, whatsapp ...

**Comment by FelixMalfait** [comment]
Yes we should definitely build that, I agree. I had logged in a similar idea in the discussion section: https://github.com/twentyhq/twenty/discussions/211

First step would be to implement a new type of activity (currently we only have notes, then tasks will come this week, a we probably need a third category that describes a past activity I want to log). Then leverage the google sign-in to import email through this as a first step. Then offering a more generic imap solution would be great.

I would say this is on the medium-term roadmap so something that is very likely to be fully implemented before end of year, but not likely to be implemented in the next 5 weeks.

**Reply by zetaron** [reply]
Even though it might be the step ahead 🙈

I've recently stumbled upon the OpenSpurce https://emailengine.app which promises easy integration of IMAP, Gmail and Outlook by use of WebHooks and REST API.

I didn't yet use it myself but it might be worth a shot.

-- 

Edit: just saw that it has a proprietary license which requires a paid license... which probably it not a good fit for a core component which might be redistributed with this project...

**Reply by GP-S** [reply]
Oh cool !I see they even have tracking for opennings and link click.

**Comment by GP-S** [comment]
For now I am using folk.app and to be honest its main main argument against twenty is how it handle email conversation. Being able to reply directly in your crm and to catch up on all onversation at a glance is a must.

**Reply by FelixMalfait** [reply]
Yes sorry we still haven't prioritized this. I can commit we'll ship email sync by EOY. Is gmail enough for you? That will be our priority.

**Reply by GP-S** [reply]
Don't worry Rome wasn't built in a day.
No we don't use gmail and probably never will, sorry.

I am just curious as it is not my field of expertise : Why priorising gmail instead of Imap sync that will open you every mailbox provider (including gmail) ?

**Reply by FelixMalfait** [reply]
Sorry, you had said it in the title 🤦‍♂️ 

I'm not 100% sure we'll prioritize gmail first but it's the most likely.
Because:
- gmail is the clear market leader amongst startups, which are the companies most likely to use Twenty
- setting up imap is easy for you but can be complex for less tech-savvy audiences ; hard to beat the simplicity of an OAuth flow

But I agree we shouldn't wait too long to introduce it as an option

**Comment by FelixMalfait** [comment]
I did a quick investigation on IMAP yesterday. I don't think it's something we want to have to maintain ourselves, it seems full of small details that would be very costly.
Instead the way we should implement this whenever someone decides to take this on is to rely on https://github.com/postalsys/emailengine and considering email engine as a Driver like Microsoft and Google are, which we will interact with through webhooks/API keys - and then email engine will take care of the back-and-fort with imap/smtp.

**Reply by zacnoo** [reply]
Is there a plan to start implementing such a service? I really like the software otherwise, but it's pretty much a dealbreaker for our company not having IMAP / SMTP support.

Is it in the plans for this year?

**Reply by nickwild-999** [reply]
LOVE the direction of Twenty, but  yes I agree this is pretty much a deal breaker not to have access to office365 emails.

**Reply by justinan7** [reply]
Agreed, once you have O365 email integration my company will be in.

**Reply by BernhardBaumrock** [reply]
@FelixMalfait any update on this? I discovered Twenty yesterday and it looks extremely promising. I'm the dev of a startup and our sales guy wants to use hubspot because he can efficiently send out mails to contacts from predefined templates/drafts.

I love open source and I love what I see so far from Twenty, but our mails are also self-hosted and thus we'd need IMAP support.

Do you have a roadmap for supporting IMAP and sending out mails? I read in the docs it's "coming soon".

Congrats on what you have built so far! Really outstanding software.

**Reply by FelixMalfait** [reply]
Hey @BernhardBaumrock IMAP support is now in beta on the main branch (need to go to Settings > Lab to enable it). Next time we do a docker version it will be in it! We also support Office 365 now with a native integration like Google.
There's an ongoing pull request for SMTP which will be merged very soon but the use-cases for outgoing emails are still relatively limited: it's only available as a node for workflow automation, there is no reply/compose button to reply to emails from the app directly at this point.

**Comment by MarcelBigger** [comment]
As e-mail is the most important business communication channel, a CRM application not showing in- and outbound e-mail messages for all clients and contacts, regardless of which e-mail server/service is used, is IMHO seriously if not fatally deficient.

Implementing native integrations for all major e-mail platforms/protocols ought to be your highest priority.

**Comment by FelixMalfait** [comment]
IMAP/SMTP/Caldav is finally complete thanks to @neo773 (it will first be available via a toggle in Settings > Lab). This will be in the upcoming docker release (end of next week), or is already available on `main` for local development. Feedback welcome!

**Reply by moan0s** [reply]
I started using it and have experienced mixed results. 
Ot's not clear to me how e-mails actually get imported. My expectation would be

- **Add an e-mail address** ✅ 
- **E-Mails are fetched from the server** ✅ I see this in the logs)
- **For each e-mail message, a person is created** ⚠️ Yes this depends on the settings, but for "Sent and received" I'd expect all messages to create a person. This does only work for sent e-mails right now
- **New e-mails are added to existing or new people** ❌ Doesn't work at all

Maybe some things are not yet implemented?

Probably this is all the same as https://github.com/twentyhq/twenty/issues/13397 and I just have to stay patient 😁

**Reply by FelixMalfait** [reply]
Thanks a lot for the beta feedback! @neo773 is looking into it and it should be fixed in the next release

**Reply by moan0s** [reply]
Sure, happy to. I'm looking forward to testing it again!

**Reply by neo773** [reply]
@moan0s 
I have opened a new PR to address these issues
Though Contact auto-creation policy should work as expected it's outside of IMAP's scope and is handled by `MessagingSaveMessagesAndEnqueueContactCreationService`

**Reply by ianaware** [reply]
@neo773 @FelixMalfait @moan0s 

I am also testing the IMAP feature. But I cannot get it to work as I expect?

I go to add another account via IMAP and enter the details of my Gmail testing account and save.  It says that the account was saved successfully, and the following screen I set the email setting (e.g. visibility and contact auto-creation settings) and I add the account and it syncs.

I then send a test email to myself and cc the IMAP account I just added to Twenty and nothing happens?  I expect myself to get added as a contact?

Here are the settings, image is zoomed out to try to fit it all in one

<img width="587" height="1146" alt="image" src="https://github.com/user-attachments/assets/84c97dc5-e5b3-4bd1-b3be-81ae7201bf86" />

Based on the contact section, I should be automatically added as a contact because I was the email account that the linked IMAP account received an email from.

It didn't work.

Instead, what I had to do was add myself as a contact in Twenty first manually and then in the Emails tab all the test emails were there.  But I shouldn't have to do this?  The email I sent should do this for me, surely?  What am I missing here?

Any help on this is much appreciated.

**Reply by neo773** [reply]
Does your handle look like a group email?  
This our regex `/noreply|no-reply|do_not_reply|no\.reply|^(info@|contact@|hello@|support@|feedback@|service@|help@|invites@|invite@|welcome@|alerts@|team@|notifications@|notification@|news@)/;`
If it does that's why it did now show up for you, I have a PR for this edge case which should be merged today https://github.com/twentyhq/twenty/pull/16508

**Reply by ianaware** [reply]
I got it to work if I emailed from my work account (MS365) to an Outlook account (my personal for testing) and CC'd the IMAP account we added to Twenty (Gmail).  

The contact was made for the Outlook account (I turned off the option to not create accounts for Outlook and Gmail accounts) and the email history was in the Person contact for the Outlook account and also for the person contact assigned to the MS365 email account that I created manually because it seems that internal emails do not get synced, hence having to use my personal Outlook for testing (originally it was my MS365 to myself and CC or BCC the Gmail IMAP) but no contact created in Twenty.

But, for my use case, we want to send emails to other colleagues in the same domain (we are all on MS365) and CC or BCC the Twenty CRM email and for it to add them to Twenty as contacts and such, but the documentation says that internal emails are not synced, why is this?  Is there the possibility that this might change?

Also, if the CRM IMAP account was added to my TwentyCRM user account, and my colleague, who also has an account with our self-hosted TwentyCRM BCCs or CCs to it using their MS365 email account, will it work for them as well and make the new contact and such if they email an account not on the same domain, same as my testing?

Furthermore, it seems that the import of email from the IMAP is not instant and is around 15 minutes or so, is that correct?

Any more help or explanation on this would be very much appreciated, thank you very much for replying @neo773, I appreciate it.
